http/agent.go

```diff
@@ -0,0 +1,273 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	defaultPostContentType = "application/octet-stream"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+// Agent is an http agent
+type Agent struct {
+	options *AgentOptions
+	AgentImplementation
+}
+
+// AgentImplementation is the actual implementation of the http calls
+//counterfeiter:generate . AgentImplementation
+type AgentImplementation interface {
+	SendPostRequest(string, []byte, *AgentOptions) (*http.Response, error)
+	SendGetRequest(string, *AgentOptions) (*http.Response, error)
+}
+
+type defaultAgentImplementation struct{}
+
+// AgentOptions has the configurable bits of the agent
+type AgentOptions struct {
+	FailOnHTTPError bool          // Set to true to fail on HTTP Status > 299
+	Timeout         time.Duration // Timeout when fetching URLs
+	Retries         int           // Number of times to retry when errors happen
+	MaxWaitTime     int           // Max waiting time when backing off
+	PostContentType string
+}
+
+// Validate checks the agent options are correct
+func (o *AgentOptions) Validate() error {
+	if o.Retries < 0 {
+		return errors.New("retries has to be equal o greater than 0")
+	}
+	logrus.Infof(
+		"HTTP.Agent options: Timeout: %d - Retries: %d - FailOnHTTPError: %+v",
+		o.Timeout, o.Retries,
+		o.FailOnHTTPError,
+	)
+	return nil
+}
+
+var DefaultAgentOptions = &AgentOptions{
+	FailOnHTTPError: true,
+	Timeout:         0,
+	Retries:         0,
+	MaxWaitTime:     60,
+	PostContentType: defaultPostContentType,
+}
+
+// NewAgent return a new agent with default options
+func NewAgent() *Agent {
+	return &Agent{
+		AgentImplementation: &defaultAgentImplementation{},
+		options:             DefaultAgentOptions,
+	}
+}
+
+// SetImplementation sets the agent implementation
+func (a *Agent) SetImplementation(impl AgentImplementation) {
+	a.AgentImplementation = impl
+}
+
+// SetOptions set the options that control the agent
+func (a *Agent) SetOptions(opts *AgentOptions) {
+	a.options = opts
+}
+
+// Options return the options controlling the agent
+func (a *Agent) Options() *AgentOptions {
+	return a.options
+}
+
+// WithTimeout returns a new agent with the specified timeout
+func WithTimeout(timeout time.Duration) *Agent {
+	return NewAgent().WithTimeout(timeout)
+}
+
+// WithTimeout sets the agent timeout
+func (a *Agent) WithTimeout(timeout time.Duration) *Agent {
+	a.Options().Timeout = timeout
+	return a
+}
+
+// WithRetries returns a new agent with the specified trries
+func WithRetries(retries int) *Agent {
+	return NewAgent().WithRetries(retries)
+}
+
+// WithRetries sets the number of times we'll attempt to fetch the URL
+func (a *Agent) WithRetries(retries int) *Agent {
+	a.Options().Retries = retries
+	return a
+}
```

Comment:
I assume that the purpose of adding a function as well as a method is for convenient usage. I would prefer that we don't do that, and simplify the API by only providing:

- A `NewAgent` function, which creates an `*Agent`, which can be already used (in theory this could be a different type `*AgentBuilder`)
- Configuration methods `With…`

Reply:
Exactly, my idea was to have a super-fast (and less verbose) way  of getting URLs, like this:

```golang
webpage:=h.Get("http://www.example.com")
webpage:=h.WithRetries(2).Get("http://www.example.com")
``` 
if we only have the `NewAgent()` function, I'll remove the Get() and Post() functions too and only leave the methods.

Reply:
OK, I've reworked the implementation functions to not require the options. Now the options type is unexported.

Reply:
OK, I've stripped all package functions and the only one remaining is `http.NewAgent()` loaded with the defaults proposed above.